### PR TITLE
Class hierarchy: add DOT output, unit tests

### PR DESCRIPTION
**Pull request by smowton**
This also adds a goto-instrument command to output the class hierarchy in text or DOT format. This is currently not usable with vanilla cbmc because none of its tools can produce GOTO binaries from Java class files, but out-of-tree tools can, and JBMC will surely gain support for GBF or its successors in the near future.

(This is another utility function from security analyser (which does use Java GBFs) which I figured we might as well upstream and save others a little time)

### CODE REVIEW

File: unit/goto-programs/class_hierarchy_output.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  std::stringstream output_stream;
+  std::stringstream output_dot_stream;
+
+  hierarchy(symbol_table);
+  hierarchy.output(output_stream);
+  hierarchy.output_dot(output_dot_stream);
+
+  std::string output = output_stream.str();
+  std::string output_dot = output_dot_stream.str();
+}
```
Comment (chrisr-diffblue) [inline unit/goto-programs/class_hierarchy_output.cpp]: I may well be going blind... but I don't think you ever call ```require_parent_child_relationship``` in this unit test, so you've never actually tested the result of the ```output*``` functions is correct.

**Review by chrisr-diffblue — CHANGES_REQUESTED** [review]
Looks nice in general, but looks like the unit test needs some work. Is it also worth adding regression tests too to ensure the command line option is working too ?

**Comment by tautschnig** [comment]
`goto-cc` should actually be able to produce goto binaries from Java bytecode, but I haven't tried this in a long time.

**Comment by smowton** [comment]
@chrisr-diffblue ahem. Oops. Fixed now.

@tautschnig it could before, but that functionality has rotted in the meantime since it isn't really supposed to be used that way. I suspect if we make a goto-javac then it should behave differently in any case -- most likely it should compile a single class file to GBF, rather than walking the class-graph as it does now, such that we end up with a .class <1-1> .class.gbf relationship. Loading multiple GBFs that were generated with goto-cc also used to fail because the linker didn't really support Java, and/or Java symbols didn't declare appropriate linker flags.

Long story short, there is no currently working way to make Java GBFs in-tree, but there probably will be in the future.

**Comment by tautschnig** [comment]
> Long story short, there is no currently working way to make Java GBFs in-tree, but there probably will be in the future.

I'm now starting to remember that @kroening and I reviewed possible ways of embedding additional information into Java bytecode (just like we do for ELF objects). That task was then superseded by more important ones...

**Comment by chrisr-diffblue** [comment]
Thanks @smowton - what is your feeling about adding a regression test (or some other sort of test) that will confirm that the command line option is recognised and works? I personally think it would be nice if we had some tests that complain if we happen to accidentally break the external UI of the tools when doing some other refactoring, etc, but I'm not sure how rigorous a convention we have around here for adding tests for command line options.

**Comment by smowton** [comment]
@chrisr-diffblue yeah, go on then. Added two regression tests that just check the options are recognised and the DOT version yields a correct graph header for a C program.

**Review by chrisr-diffblue — APPROVED** [review]
Thanks @smowton. Looks good to me.

**Review by tautschnig — APPROVED** [review]
